# Shelter Animals

Homework 1
Camille Girard


You are volunteering at an animal shelter with cats or dogs in the shelter at a given time. A family comes in to look at 3 dogs and 5 cats. What are they most likely to do (Assuming there is an equal chance to pick each pet)?

How different is a family that looks at 10 dogs and 1 cat? 1 dog and 10 cats?

(Adopted or not? Data from the ASPCA - Average shelters contain 51% dogs and 49% cats. Dogs are 0.68 adopted or returned to owners 0.32 aren't, cats 0.53 are and 0.47 aren't)

# Solution

In [102]:
# Configure Jupyter so figures appear in the notebook
%matplotlib inline

# Configure Jupyter to display the assigned value after an assignment
%config InteractiveShell.ast_node_interactivity='last_expr_or_assign'

# import classes from thinkbayes2
from thinkbayes2 import Pmf, Suite

import thinkplot
import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings('ignore')

ERROR:root:The 'ast_node_interactivity' trait of a ZMQInteractiveShell instance must be any of ['all', 'last', 'last_expr', 'none'], but a value of 'last_expr_or_assign' <class 'str'> was specified.


In [116]:
#Create a table for the computations
class BayesTable(pd.DataFrame):
    def __init__(self, hypo, prior=1):
        #create the table with the initial sections
        columns = ['hypo', 'prior', 'likelihood', 'unnorm', 'posterior']
        super().__init__(columns=columns)
        self.hypo = hypo
        self.prior = prior
    
    def mult(self):
        #set a function to multiply the prior by the likelihood
        self.unnorm = self.prior * self.likelihood
        
    def norm(self):
        #set a function to normalize the values
        nc = np.sum(self.unnorm)
        self.posterior = self.unnorm / nc
        return nc
    
    def update(self):
        #set a function that will do both mult and norm in one step
        self.mult()
        return self.norm()
    
    def reset(self):
        return BayesTable(self.hypo, self.posterior)

In [117]:
# create a table with the priors
# the priors are the portion of that animal in the shelter multiplied by the chance that it is adopted or not
table2 = BayesTable(['Dog-Adopt', 'Dog-Not', 'Cat-Adopt', 'Cat-Not'], [0.51*0.68,0.51*0.32, 0.49*0.53, 0.49*0.37])
#create a likelihood that shows how many dogs and cats the family has looked at - equal chance for each
table2.likelihood =[3/8,3/8,5/8,5/8]
table2.update()
table2

hypo   prior  likelihood    unnorm  posterior
0  Dog-Adopt  0.3468       0.375  0.130050   0.278554
1    Dog-Not  0.1632       0.375  0.061200   0.131084
2  Cat-Adopt  0.2597       0.625  0.162312   0.347657
3    Cat-Not  0.1813       0.625  0.113312   0.242704

This result tells us that they are most likely (35%) to adopt a cat and about 63% likely to adopt either a dog or cat. They still have a 37% chance of not adopting any of the pets.

In [119]:
table3 = table2.reset()
# create a new table with the same priors but new likelihood based on the animals they have looked at
table3.likelihood = [10/11,10/11,1/11,1/11]
table3.update()
table3

hypo     prior  likelihood    unnorm  posterior
0  Dog-Adopt  0.278554    0.909091  0.253231   0.594344
1    Dog-Not  0.131084    0.909091  0.119168   0.279692
2  Cat-Adopt  0.347657    0.090909  0.031605   0.074179
3    Cat-Not  0.242704    0.090909  0.022064   0.051785

This family is 59% likely to adopt a dog and only another 7% to adopt a cat, they have a smaller (34%) chance of not adopting an pets.

In [115]:
table4 = table2.reset()
table4.likelihood = [1/11,1/11,10/11,10/11]
table4.update()
table4

hypo     prior  likelihood    unnorm  posterior
0  Dog-Adopt  0.278554    0.090909  0.025323   0.044122
1    Dog-Not  0.131084    0.090909  0.011917   0.020763
2  Cat-Adopt  0.347657    0.909091  0.316052   0.550679
3    Cat-Not  0.242704    0.909091  0.220640   0.384436

This final family is 55% likely to adopt a cat however have the highest chance (40%) of not adopting any pets.